forest/cli/src/cli/snapshot_cmd.rs

```diff
@@ -323,6 +349,128 @@ fn clean(config: &Config, snapshot_dir: &Option<PathBuf>, force: bool) -> anyhow
     Ok(())
 }
 
+async fn validate(
+    config: &Config,
+    recent_stateroots: &i64,
+    snapshot: &PathBuf,
+) -> anyhow::Result<()> {
+    let confirm = Confirm::with_theme(&ColorfulTheme::default())
+        .with_prompt(format!(
+            "This will result in using approximately {} MB of data. Proceed?",
+            std::fs::metadata(snapshot)?.len() / (1024 * 1024)
+        ))
+        .default(false)
+        .interact()
+        .unwrap_or_default();
+
+    if confirm {
+        let tmp_db_path = TempDir::new()?;
+
+        let db_path = tmp_db_path.path().join(&config.chain.name);
+
+        let db = forest_db::rocks::RocksDb::open(db_path, &config.rocks_db)
+            .expect("Opening RocksDB must succeed");
+
+        let chain_store = Arc::new(ChainStore::new(db).await);
+
+        let genesis = read_genesis_header(
+            config.client.genesis_file.as_ref(),
+            config.chain.genesis_bytes(),
+            &chain_store,
+        )
+        .await
+        .unwrap();
+        let genesis_header = &genesis.blocks()[0];
+        chain_store.set_genesis(genesis_header).unwrap();
+
+        let cids = {
+            let file = tokio::fs::File::open(&snapshot).await?;
+            let reader = FetchProgress::fetch_from_file(file).await?;
+            load_car(chain_store.blockstore(), reader.compat()).await?
+        };
+
+        let ts = chain_store.tipset_from_keys(&TipsetKeys::new(cids)).await?;
+
+        validate_links_and_genesis_traversal(
+            &chain_store,
+            ts,
+            chain_store.blockstore(),
+            *recent_stateroots,
+            &genesis,
+        )
+        .await?;
+    }
+
+    Ok(())
+}
+
+async fn validate_links_and_genesis_traversal<DB>(
+    chain_store: &ChainStore<DB>,
+    ts: Arc<Tipset>,
+    db: &DB,
+    recent_stateroots: ChainEpoch,
+    genesis_tipset: &Tipset,
+) -> anyhow::Result<()>
+where
+    DB: fvm_ipld_blockstore::Blockstore + Store + Send + Sync,
+{
+    let mut seen = std::collections::HashSet::<Cid>::new();
+    let upto = ts.epoch() - recent_stateroots;
+
+    println!(
+        "validating cids from snapshot in db: {} <- {} ...",
+        upto,
+        ts.epoch()
+    );
```

Comment:
How about we use a progress bar instead of this message?

Reply:
added